test/testHelpers.hpp

```diff
@@ -979,4 +996,231 @@ ::testing::AssertionResult assertArrayNear(std::string hA_name, std::string aDim
+                        // the af_* function, and thus what will be released
+                        // after it is used
+    af::index out_subarr_idxs[4];
+    af::array subarr;
+    TestOutputArrayType out_arr_type;
+
+public:
+    // There might be a better way of encapsulating these, but for now, these
+    // are directly exposed to simplify dealing with af::array reference counts
+    af::array out_arr;
+    af::array out_arr_cpy;
+
+    TestOutputArrayInfo(TestOutputArrayType arr_type)
+        :out_arr_type(arr_type) {}
+
+    ~TestOutputArrayInfo() {
+        if (out_arr_c != 0) af_release_array(out_arr_c);
+    }
+
+    TestOutputArrayType getOutputArrayType() { return out_arr_type; }
+
+    void initArrays() {
+        out_arr_c = 0;
+    }
+
+    void initArrays(af::dim4 dims, af::dtype ty) {
+        out_arr = af::randu(dims, ty);
+        out_arr_cpy = out_arr.copy();
+        af_retain_array(&out_arr_c, out_arr.get());
+    }
+
+    void initArrays(af::dim4 dims, af::dtype ty, af::index subarr_idxs[4]) {
+        out_arr = af::randu(dims, ty);
+        out_arr_cpy = out_arr.copy();
+
+        for (int i = 0; i < 4; ++i) {
+            out_subarr_idxs[i] = subarr_idxs[i];
+        }
+        subarr = out_arr(
+            out_subarr_idxs[0],
+            out_subarr_idxs[1],
+            out_subarr_idxs[2],
+            out_subarr_idxs[3]);
+
+        // The output array that will be given to the client is the subarray,
+        // not the full array
+        af_retain_array(&out_arr_c, subarr.get());
+    }
+
+    af_array getOutputArrayC() {
+        return out_arr_c;
+    }
+
+    void setOutputArray(af::array array) {
+        // When reassigning the output array, release the old retained af_array
+        // then retain the new af_array of the new C++ array
+        af_release_array(out_arr_c);
+        out_arr = array;
+        af_retain_array(&out_arr_c, out_arr.get());
+    }
+
+    // This is used in the case of NULL_ARRAY, when taking the output array
+    // generated by the af_* function. Doing this gives TestOutputArrayInfo the
+    // responsibility to release that output array
+    void setOutputArrayC(af_array array) {
+        out_arr_c = array;
+    }
+
+    af::index getSubArrayIdx(dim_t dim) {
+        return out_subarr_idxs[dim];
+    }
+};
+
+// Generates a null af_array. testWriteToOutputArray expects that the array that
+// it receives after the af_* function is a valid, allocated af_array
+void genNullArray(TestOutputArrayInfo& metadata, const af::dim4& dims,
+                  const af::dtype ty) {
+    metadata.initArrays();
+}
+
+// Generates a random array. testWriteToOutputArray expects that it will receive
+// the same af_array that this generates after the af_* function is called
+void genRegularArray(TestOutputArrayInfo& metadata, const af::dim4& dims,
+                  const af::dtype ty) {
+    metadata.initArrays(dims, ty);
```

Comment:
Will become `metadata.init(dims, ty)`

Reply:
`metadata.init(dims, ty)`, according to your class definition above, simply creates a random `af::array` and stores it within the class. I think that might be too simple for this reason: We do store the `af::array` as a member data of the class, but we don't store the final `af_array` value that we give back to the test for `af_approx1` to use as its output. We need to store that `af_array` value because we need to test if `af_approx1` (or any other `af_*` function) will place its output data on the the same existing, allocated `af_array` that we give to it, and hence, we need to compare the `af_array` value we generated and the `af_array` that `af_approx1` gives as its output - they must be the same.